SOAL NO 1

In [34]:
import pandas as pd
import ssl
from sklearn import model_selection
from sklearn.tree import DecisionTreeClassifier
from sklearn.metrics import accuracy_score, classification_report
from sklearn.preprocessing import LabelEncoder

# Disable SSL certificate verification
ssl._create_default_https_context = ssl._create_unverified_context

url = "https://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']
dataset = pd.read_csv(url, names=names)

# Gunakan LabelEncoder untuk mengonversi label kelas string menjadi label numerik
le = LabelEncoder()
dataset['class'] = le.fit_transform(dataset['class'])

# Split-out validation dataset
array = dataset.values
X = array[:, 0:4]
Y = array[:, 4]  # Menggunakan kolom terakhir yang sekarang adalah label numerik
validation_size = 0.20
seed = 7
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

# Membuat model Decision Tree
dt_model = DecisionTreeClassifier()

# Melatih model pada dataset pelatihan
dt_model.fit(X_train, Y_train)

# Menguji model pada dataset pengujian
Y_pred = dt_model.predict(X_validation)

# Mengukur performa model
accuracy = accuracy_score(Y_validation, Y_pred)
print("Accuracy:", accuracy)

# Menampilkan laporan klasifikasi
print("\nClassification Report:\n", classification_report(Y_validation, Y_pred))


Accuracy: 0.8666666666666667

Classification Report:
               precision    recall  f1-score   support

         0.0       1.00      1.00      1.00         7
         1.0       0.83      0.83      0.83        12
         2.0       0.82      0.82      0.82        11

    accuracy                           0.87        30
   macro avg       0.88      0.88      0.88        30
weighted avg       0.87      0.87      0.87        30



PENJELASAN NO 1:

KODE:

-pandas digunakan untuk manipulasi dan analisis data.

-ssl digunakan untuk menangani sertifikat SSL (dalam hal ini, untuk menonaktifkan verifikasi sertifikat).

-model_selection menyediakan fungsi untuk pembagian dataset.

-DecisionTreeClassifier adalah kelas dari scikit-learn untuk membangun model pohon keputusan.

-accuracy_score dan classification_report digunakan untuk mengevaluasi performa model.

-LabelEncoder digunakan untuk mengubah label kelas string menjadi label numerik.

-Pertama, menggunakan LabelEncoder untuk mengonversi label kelas string menjadi label numerik. Ini dilakukan agar model dapat memahami label kelas sebagai nilai numerik.

-X: Fitur dari dataset Iris (kolom 0 hingga 3).

-Y: Label dari dataset Iris (kolom 4).

-test_size: Menentukan ukuran dataset validasi sebagai pecahan dari keseluruhan dataset. Dalam kasus ini, validation_size diatur ke 0.20, yang berarti 20% dari data akan digunakan sebagai dataset validasi.

-random_state: Digunakan untuk menghasilkan hasil yang konsisten setiap kali kode dijalankan. Ini berguna agar pembagian dataset dapat direproduksi dengan cara yang sama.

-X_train: Fitur dari dataset yang akan digunakan untuk melatih model.

-X_validation: Fitur dari dataset yang akan digunakan untuk menguji model.

-Y_train: Label yang sesuai dengan X_train.

-Y_validation: Label yang sesuai dengan X_validation.

-Kedua, membuat model Decision Tree dan melatihnya menggunakan data latih.

-Ketiga, menggunakan model yang telah dilatih untuk membuat prediksi pada data validasi dan mengukur akurasi prediksi.

-Terakhir, menampilkan akurasi model dan laporan klasifikasi yang mencakup precision, recall, dan f1-score untuk setiap kelas, serta nilai rata-rata berponderasi.

OUTPUT:

-akurasi sekitar 0.87 atau 87%, yang berarti model mampu memprediksi kelas dengan benar untuk sekitar 87% dari data validasi.

-Model ini memiliki akurasi yang baik, terutama mengenai kelas 0 yang memiliki precision dan recall 1.00, mencapai f1-score 1.00.

-Untuk kelas 1 dan 2, model juga memberikan hasil yang baik, dengan precision dan recall di atas 0.80.

-Macro avg dan weighted avg memberikan gambaran umum tentang performa model secara keseluruhan, dan keduanya mencerminkan akurasi sekitar 0.87.

KESIMPULAN:

Hasil ini menunjukkan bahwa model Decision Tree yang telah dilatih dengan dataset tersebut memiliki kinerja yang baik dalam memprediksi kelas pada dataset validasi. Namun, evaluasi lebih lanjut atau fine-tuning mungkin diperlukan tergantung pada kebutuhan spesifik dan tujuan analisis.

SOAL NO 2

In [3]:
import pandas as pd
from sklearn import model_selection
from sklearn.preprocessing import LabelEncoder
from sklearn.neighbors import KNeighborsClassifier
from sklearn.naive_bayes import GaussianNB
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, classification_report, confusion_matrix

url = "https://archive.ics.uci.edu/static/public/50/data.csv"
names = ['class', 'region-centroid-col', 'region-centroid-row', 'region-pixel-count', 'short-line-density-5',
         'short-line-density-2', 'vedge-mean', 'vedge-sd', 'hedge-mean', 'hedge-sd', 'intensity-mean', 'rawred-mean',
         'rawblue-mean', 'rawgreen-mean', 'exred-mean', 'exblue-mean', 'exgreen-mean', 'value-mean', 'saturation-mean',
         'hue-mean']
dataset = pd.read_csv(url, names=names, skiprows=1)  # skip the first row

# Gunakan LabelEncoder untuk mengonversi label kelas string menjadi label numerik
le = LabelEncoder()
dataset['class'] = le.fit_transform(dataset['class'])

# Split-out validation dataset
array = dataset.values
X = array[:, 1:]  # Fitur dari kolom kedua hingga terakhir
Y = array[:, 0]   # Label (class) dari kolom pertama
validation_size = 0.20
seed = 7
X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size,random_state=seed)

# Test options and Evaluasi Perhitungan
seed = 7
scoring = 'accuracy'

models = []
models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))

# Evaluasi setiap model
results = []
names = []
for name, model in models:
    kfold = model_selection.KFold(n_splits=10, random_state=seed, shuffle=True)
    cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring, error_score='raise')
    results.append(cv_results)
    names.append(name)
    msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
    print(msg)

# Membuat Prediksi pada dataset validasi
nb = SVC()
nb.fit(X_train, Y_train)
predictions = nb.predict(X_validation)
print(accuracy_score(Y_validation, predictions))
print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))


KNN: 0.768750 (0.107946)
NB: 0.786397 (0.099012)
SVM: 0.744118 (0.092880)
0.7380952380952381
[[ 5  0  0  0  0  0  0]
 [ 1 10  0  0  0  0  1]
 [ 7  0  0  0  0  0  2]
 [ 0  0  0  5  0  0  0]
 [ 0  0  0  0  7  0  0]
 [ 0  0  0  0  0  2  0]
 [ 0  0  0  0  0  0  2]]
              precision    recall  f1-score   support

         0.0       0.38      1.00      0.56         5
         1.0       1.00      0.83      0.91        12
         2.0       0.00      0.00      0.00         9
         3.0       1.00      1.00      1.00         5
         4.0       1.00      1.00      1.00         7
         5.0       1.00      1.00      1.00         2
         6.0       0.40      1.00      0.57         2

    accuracy                           0.74        42
   macro avg       0.68      0.83      0.72        42
weighted avg       0.68      0.74      0.69        42



/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))
/usr/local/lib/python3.10/dist-packages/sklearn/metrics/_classification.py:1344: UndefinedMetricWarning: Precision and F-score are ill-defined and being set to 0.0 in labels with no predicted samples. Use `zero_division` parameter to control this behavior.
  _warn_prf(average, modifier, msg_start, len(result))


PENJELASAN NO 2:

KODE:

-Pertama, menggunakan LabelEncoder untuk mengonversi label kelas dari string menjadi label numerik.

--X: Fitur dari dataset Image Segmentation (kolom 1 dan seterusnya).

-Y: Label dari dataset Image Segmentation (kolom 0).

-test_size: Menentukan ukuran dataset validasi sebagai pecahan dari keseluruhan dataset. Dalam kasus ini, validation_size diatur ke 0.20, yang berarti 20% dari data akan digunakan sebagai dataset validasi.

-random_state: Digunakan untuk menghasilkan hasil yang konsisten setiap kali kode dijalankan. Ini berguna agar pembagian dataset dapat direproduksi dengan cara yang sama.

-X_train: Fitur dari dataset yang akan digunakan untuk melatih model.

-X_validation: Fitur dari dataset yang akan digunakan untuk menguji model.

-Y_train: Label yang sesuai dengan X_train.

-Y_validation: Label yang sesuai dengan X_validation.

-Kedua, melatih dan mengevaluasi tiga model machine learning (KNN, NB, SVM) menggunakan cross-validation.

-Ketiga, menampilkan hasil evaluasi untuk setiap model, termasuk mean dan standard deviation dari akurasi.

-menggunakan K-Fold Cross-Validation dengan 10 lipatan (n_splits=10).

-random_state digunakan agar pembagian lipatan dapat direproduksi dengan cara yang sama.

-shuffle=True mengacak dataset sebelum pembagian lipatan.

-menggunakan cross_val_score untuk melakukan cross-validation pada model tertentu (model) dengan data latih (X_train, Y_train).
scoring ditentukan sebelumnya sebagai 'accuracy', yang akan menghitung akurasi pada setiap lipatan.

-Keempat, membuat model SVM baru (nb), melatihnya dengan data latih, dan melakukan prediksi pada data validasi.

-Terakhir, menampilkan akurasi, confusion matrix, dan laporan klasifikasi untuk hasil prediksi pada data validasi.

OUTPUT:

-KNN: Rata-rata akurasi sekitar 0.768 dengan deviasi standar sekitar 0.108.

-NB: Rata-rata akurasi sekitar 0.786 dengan deviasi standar sekitar 0.099.

-SVM: Rata-rata akurasi sekitar 0.744 dengan deviasi standar sekitar 0.093.

-Hasil di atas merupakan hasil rata-rata akurasi dan deviasi standar dari masing-masing model selama proses cross-validation dengan 10 lipatan. Rata-rata akurasi memberikan gambaran tentang seberapa baik model dapat memprediksi pada data latih, sedangkan deviasi standar memberikan informasi tentang seberapa konsisten model dalam melakukan prediksi.

-Akurasi pada data validasi adalah sekitar 0.738, yang berarti model dapat memprediksi dengan benar untuk sekitar 73.8% dari data validasi.

-Confusion matrix memberikan informasi tentang seberapa baik model dalam mengklasifikasikan setiap kelas. Misalnya, kelas 2 tidak memiliki prediksi yang benar.

-Classification report memberikan informasi lebih lanjut tentang precision, recall, dan f1-score untuk setiap kelas, serta rata-rata berponderasi.